File: src/applications/baseConverter.ts
import { StackWithObject } from "@lib/stack"

export function baseConverter(decimalNumber: number, base: number): string {
    if (base < 2 || base > 36) {
        return ""
    }

    const remStack = new StackWithObject<number>()
    const digits = "0123456789ABCDEFGHYJKLMNOPQRSTUVWXYZ"

    let num = decimalNumber
    let rem = 0;
    let baseString = "";

    while (num > 0) {
        rem = Math.floor(num % base);
        remStack.push(rem)
        num = Math.floor(num / base);
    }

    while (!remStack.isEmpty()) {
        // @ts-ignore
        baseString += digits[remStack.pop()]
    }

    return baseString
}

File: lua/applications/base_converter.lua
require("math")
require("lib/stack")

--- comment convert a number to another base
--- @param decimalNumber number
--- @param base number
function BaseConverter(decimalNumber, base)
    if base < 2 or base > 36 then
        return ""
    end

    local stack = Stack.new()
    local digits = "0123456789ABCDEFGHYJKLMNOPQRSTUVWXYZ"

    local num = decimalNumber
    local rem = 0
    local baseString = ""

    while num > 0 do
        rem = math.floor(num % base)
        stack:push(rem)
        num = math.floor(num / base)
    end

    while not stack:is_empty() do
        local pop = stack:pop() + 1
        if pop ~= nil then
            baseString = baseString .. digits:sub(pop, pop)
        end
    end

    return baseString
end

print(BaseConverter(255, 16))
print(BaseConverter(10, 2))
print(BaseConverter(31, 8))
print(BaseConverter(2, 2))
